File: Beecrowd_Parte1/A_Paz_em_Sparta_2860.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Edge {
    int dest;
    int weight;

    public Edge(int dest, int weight) {
        this.dest = dest;
        this.weight = weight;
    }
}

class Graph {
    int V;
    List<Edge>[] adjList;

    public Graph(int V) {
        this.V = V;
        adjList = new ArrayList[V];
        for (int i = 0; i < V; i++) {
            adjList[i] = new ArrayList<>();
        }
    }

    public void addEdge(int src, int dest, int weight) {
        adjList[src].add(new Edge(dest, weight));
        adjList[dest].add(new Edge(src, weight));
    }

    public long primMST() {
        long[] key = new long[V]; // Para armazenar as chaves de seleção
        boolean[] mstSet = new boolean[V]; // Para representar os vértices incluídos na AGM
        Arrays.fill(key, Long.MAX_VALUE);

        // Min-Heap usando PriorityQueue
        PriorityQueue<Edge> pq = new PriorityQueue<>(V, Comparator.comparingLong(e -> e.weight));

        // Começando com o vértice 0
        pq.add(new Edge(0, 0));
        key[0] = 0;

        while (!pq.isEmpty()) {
            int u = pq.poll().dest;

            mstSet[u] = true;

            for (Edge e : adjList[u]) {
                int v = e.dest;
                int weight = e.weight;

                if (!mstSet[v] && weight < key[v]) {
                    key[v] = weight;
                    pq.add(new Edge(v, key[v]));
                }
            }
        }

        // Encontra a maior distância na AGM
        long maxDistance = 0;
        for (int i = 0; i < V; i++) {
            maxDistance += key[i];
        }

        return maxDistance;
    }
}

public class A_Paz_em_Spar {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim()); // Número de casos de teste

        StringBuilder sb = new StringBuilder();

        for (int caseNum = 0; caseNum < t; caseNum++) {
            int n = Integer.parseInt(br.readLine().trim()); // Número de cidades
            Graph graph = new Graph(n);

            // Lê as arestas do grafo
            for (int i = 0; i < n - 1; i++) {
                String[] edgeInfo = br.readLine().trim().split(" ");
                int a = Integer.parseInt(edgeInfo[0]);
                int b = Integer.parseInt(edgeInfo[1]);
                graph.addEdge(a, b, 1); // Todas as arestas têm peso 1
            }

            int q = Integer.parseInt(br.readLine().trim()); // Número de novas estradas a serem consideradas

            // Para cada nova estrada
            for (int i = 0; i < q; i++) {
                String[] newRoadInfo = br.readLine().trim().split(" ");
                int u = Integer.parseInt(newRoadInfo[0]);
                int v = Integer.parseInt(newRoadInfo[1]);

                // Adiciona a nova estrada ao grafo temporariamente
                graph.addEdge(u, v, 1);

                // Calcula a Árvore Geradora Mínima (AGM) e a maior distância
                long maxDistance = graph.primMST();

                // Remove a nova estrada do grafo para o próximo cálculo
                graph.adjList[u].removeIf(e -> e.dest == v);
                graph.adjList[v].removeIf(e -> e.dest == u);

                // Adiciona a resposta ao StringBuilder
                sb.append(maxDistance).append("\n");
            }

            if (caseNum < t - 1) {
                sb.append("\n"); // Linha em branco entre os casos de teste
            }
        }

        System.out.print(sb.toString());
    }
}

File: Beecrowd_Parte1/A_Paz_em_Spar-Ta_2860.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <limits.h>

// Estrutura para representar um nó da lista de adjacências
typedef struct Node {
    int dest;
    int weight;
    struct Node* next;
} Node;

// Estrutura para representar um grafo
typedef struct Graph {
    int V; // Número de vértices
    Node** adjList; // Lista de adjacências
} Graph;

// Função para criar um novo nó da lista de adjacências
Node* createNode(int dest, int weight) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->dest = dest;
    newNode->weight = weight;
    newNode->next = NULL;
    return newNode;
}

// Função para criar um grafo com V vértices
Graph* createGraph(int V) {
    Graph* graph = (Graph*)malloc(sizeof(Graph));
    graph->V = V;
    graph->adjList = (Node**)malloc(V * sizeof(Node*));
    
    for (int i = 0; i < V; ++i)
        graph->adjList[i] = NULL;
    
    return graph;
}

// Função para adicionar uma aresta não direcionada ao grafo
void addEdge(Graph* graph, int src, int dest, int weight) {
    // Adiciona aresta de src para dest
    Node* newNode = createNode(dest, weight);
    newNode->next = graph->adjList[src];
    graph->adjList[src] = newNode;

    // Adiciona aresta de dest para src
    newNode = createNode(src, weight);
    newNode->next = graph->adjList[dest];
    graph->adjList[dest] = newNode;
}

// Função para liberar a memória alocada para o grafo
void freeGraph(Graph* graph) {
    for (int i = 0; i < graph->V; ++i) {
        Node* current = graph->adjList[i];
        while (current != NULL) {
            Node* temp = current;
            current = current->next;
            free(temp);
        }
    }
    free(graph->adjList);
    free(graph);
}

// Função auxiliar para o algoritmo de Prim
int minKey(int key[], bool mstSet[], int V) {
    int min = INT_MAX, min_index;

    for (int v = 0; v < V; v++) {
        if (mstSet[v] == false && key[v] < min) {
            min = key[v];
            min_index = v;
        }
    }
    return min_index;
}

// Algoritmo de Prim para encontrar a Árvore Geradora Mínima (AGM)
long long primMST(Graph* graph) {
    int V = graph->V;
    int parent[V]; // Array para armazenar a árvore geradora mínima
    int key[V];    // Chaves usadas para pegar o peso mínimo de corte

    bool mstSet[V]; // Para representar os vértices incluídos na árvore geradora mínima

    // Inicializa todas as chaves como INFINITO
    // e mstSet[] como falso
    for (int i = 0; i < V; i++) {
        key[i] = INT_MAX;
        mstSet[i] = false;
    }

    // A chave do primeiro vértice é sempre 0
    key[0] = 0;     // Fazemos o primeiro vértice como raiz da AGM
    parent[0] = -1; // Primeiro vértice não tem pai

    // O grafo contém V vértices, então existem V-1 arestas para adicionar à AGM
    for (int count = 0; count < V - 1; count++) {
        // Escolhe a chave mínima do vértice ainda não incluído
        int u = minKey(key, mstSet, V);

        // Inclui o vértice escolhido na árvore geradora mínima
        mstSet[u] = true;

        // Atualiza as chaves e os pais dos vértices adjacentes do vértice escolhido
        Node* iter = graph->adjList[u];
        while (iter != NULL) {
            int v = iter->dest;
            int weight = iter->weight;

            if (mstSet[v] == false && weight < key[v]) {
                parent[v] = u;
                key[v] = weight;
            }
            iter = iter->next;
        }
    }

    // Calcula a maior distância na AGM entre qualquer vértice e seu representante mais próximo
    long long maxDistance = 0;
    for (int i = 1; i < V; i++) {
        maxDistance += key[i];
    }

    return maxDistance;
}

int main() {
    int t; // Quantidade de casos de teste
    scanf("%d", &t);

    while (t--) {
        int n; // Número de cidades
        scanf("%d", &n);

        // Cria o grafo com n vértices
        Graph* graph = createGraph(n);

        // Leitura das arestas do grafo
        for (int i = 0; i < n - 1; i++) {
            int a, b;
            scanf("%d %d", &a, &b);
            addEdge(graph, a, b, 1); // Todas as arestas têm peso 1
        }

        int q; // Número de novas estradas a serem consideradas
        scanf("%d", &q);

        // Para cada nova estrada
        for (int i = 0; i < q; i++) {
            int u, v;
            scanf("%d %d", &u, &v);

            // Adiciona temporariamente a nova aresta ao grafo
            addEdge(graph, u, v, 1);

            // Calcula a Árvore Geradora Mínima (AGM) e a maior distância
            long long maxDistance = primMST(graph);

            // Remove a nova aresta do grafo para o próximo cálculo
            Node* iter = graph->adjList[u];
            while (iter != NULL) {
                if (iter->dest == v) {
                    Node* temp = iter;
                    iter = iter->next;
                    free(temp);
                    graph->adjList[u] = iter;
                    break;
                }
                iter = iter->next;
            }

            iter = graph->adjList[v];
            while (iter != NULL) {
                if (iter->dest == u) {
                    Node* temp = iter;
                    iter = iter->next;
                    free(temp);
                    graph->adjList[v] = iter;
                    break;
                }
                iter = iter->next;
            }

            // Imprime a maior distância calculada
            printf("%lld\n", maxDistance);
        }

        // Libera o grafo após cada caso de teste
        freeGraph(graph);

        if (t > 0) {
            printf("\n"); // Linha em branco entre os casos de teste
        }
    }

    return 0;
}
